storage/Devices/MdMemberImpl.cc

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2017 SUSE LLC
+ *
+ * All Rights Reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of version 2 of the GNU General Public License as published
+ * by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact Novell, Inc.
+ *
+ * To contact Novell about this file by physical or electronic mail, you may
+ * find current contact information at www.novell.com.
+ */
+
+
+#include "storage/Devices/MdMemberImpl.h"
+#include "storage/Devices/MdContainerImpl.h"
+#include "storage/Devices/MdImpl.h"
+#include "storage/SystemInfo/SystemInfo.h"
+#include "storage/Holders/MdSubdeviceImpl.h"
+#include "storage/EtcMdadm.h"
+#include "storage/Prober.h"
+#include "storage/Utils/StorageTmpl.h"
+
+
+namespace storage
+{
+
+    using namespace std;
+
+
+    const char* DeviceTraits<MdMember>::classname = "MdMember";
+
+
+    MdMember::Impl::Impl(const string& name)
+	: Md::Impl(name)
+    {
+    }
+
+
+    MdMember::Impl::Impl(const xmlNode* node)
+	: Md::Impl(node)
+    {
+    }
+
+
+    void
+    MdMember::Impl::check() const
+    {
+	Md::Impl::check();
+
+	if (get_metadata() != "")
+	    ST_THROW(Exception("invalid metadata for MdMember"));
+
+	// TODO check that container exists
+    }
+
+
+    MdContainer*
+    MdMember::Impl::get_md_container()
+    {
+	vector<MdContainer*> tmp = get_parents_of_type<MdContainer>();
+	if (tmp.size() != 1)
+	    ST_THROW(Exception("MdMember has no MdContainer"));
+
+	return tmp[0];
+    }
+
+
+    const MdContainer*
+    MdMember::Impl::get_md_container() const
+    {
+	vector<const MdContainer*> tmp = get_parents_of_type<const MdContainer>();
+	if (tmp.size() != 1)
+	    ST_THROW(Exception("MdMember has no MdContainer"));
+
+	return tmp[0];
+    }
+
+
+    vector<BlkDevice*>
+    MdMember::Impl::get_devices()
+    {
+	vector<BlkDevice*> tmp = Md::Impl::get_devices();
+
+	erase_if(tmp, [](BlkDevice* blk_device) { return is_md_container(blk_device); });
+
+	return tmp;
+    }
+
+
+    vector<const BlkDevice*>
+    MdMember::Impl::get_devices() const
+    {
+	vector<const BlkDevice*> tmp = Md::Impl::get_devices();
+
+	erase_if(tmp, [](const BlkDevice* blk_device) { return is_md_container(blk_device); });
```

Comment:
The parent for a MdMember is the MdContainer and the BlkDevices the MdMember uses. MdMember::get_devices() returns only the BlkDevices, MdMember::get_md_container() returns only the MdContainer.